pkg/restore/cdclog/buffer.go

```diff
@@ -0,0 +1,165 @@
+// this is the concurrent unit of log restore.
+type TableBuffer struct {
+	KvPairs []Row
+	Size    int64
+
+	KvEncoder Encoder
+	tableInfo table.Table
+
+	flushKVPairs int
+
+	colNames []string
+	colPerm  []int
+}
+
+// NewTableBuffer creates TableBuffer.
+func NewTableBuffer(tbl table.Table, flushKVPairs int) *TableBuffer {
+	kvEncoder := NewTableKVEncoder(tbl, &SessionOptions{
+		Timestamp: time.Now().Unix(),
+		// TODO make it config
+		RowFormatVersion: "1",
+	})
+
+	columns := tbl.Meta().Cols()
+	colNames := make([]string, 0, len(columns))
+	colPerm := make([]int, 0, len(columns)+1)
+
+	for i, col := range columns {
+		colNames = append(colNames, col.Name.String())
+		colPerm = append(colPerm, i)
+	}
+	if TableHasAutoRowID(tbl.Meta()) {
+		colPerm = append(colPerm, -1)
+	}
+
+	return &TableBuffer{
+		KvPairs:   make([]Row, 0, flushKVPairs),
+		KvEncoder: kvEncoder,
+		tableInfo: tbl,
+
+		flushKVPairs: flushKVPairs,
+
+		colNames: colNames,
+		colPerm:  colPerm,
+	}
+}
+
+func (t *TableBuffer) translateToDatum(row map[string]column) ([]types.Datum, error) {
+	cols := make([]types.Datum, 0, len(row))
+	for _, col := range t.colNames {
+		val, err := row[col].toDatum()
+		if err != nil {
+			return nil, err
+		}
+		log.Debug("translate to datum", zap.String("col", col), zap.Stringer("val", val))
+		cols = append(cols, val)
+	}
+	return cols, nil
+}
+
+// Append appends the item to this buffer.
+func (t *TableBuffer) Append(ctx context.Context, item *SortItem) error {
+	log.Debug("Append item to buffer",
+		zap.Stringer("table", t.tableInfo.Meta().Name),
+	)
+	row := item.Meta.(*MessageRow)
+
+	if row.Update != nil {
+		if row.PreColumns != nil {
+			log.Debug("process update event", zap.Any("row", row))
+			oldCols, err := t.translateToDatum(row.PreColumns)
+			if err != nil {
+				return err
+			}
+
+			pair, err := t.KvEncoder.RemoveRecord(oldCols, item.RowID, t.colPerm)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			t.KvPairs = append(t.KvPairs, pair)
+			t.Size++
+
+			newCols, err := t.translateToDatum(row.Update)
+			if err != nil {
+				return err
+			}
+			pair, err = t.KvEncoder.AddRecord(newCols, item.RowID, t.colPerm)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			t.KvPairs = append(t.KvPairs, pair)
+			t.Size++
+		} else {
+			log.Debug("process insert event", zap.Any("row", row))
+			cols, err := t.translateToDatum(row.Update)
+			if err != nil {
+				return err
+			}
+			pair, err := t.KvEncoder.AddRecord(cols, item.RowID, t.colPerm)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			t.KvPairs = append(t.KvPairs, pair)
+			t.Size++
+		}
+	}
+	if row.Delete != nil {
+		log.Debug("process delete event", zap.Any("row", row))
+		cols, err := t.translateToDatum(row.Delete)
+		if err != nil {
+			return err
+		}
+		pair, err := t.KvEncoder.RemoveRecord(cols, item.RowID, t.colPerm)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		t.KvPairs = append(t.KvPairs, pair)
+		t.Size++
```

Comment:
seems these codes in insert/update/delete event are almost the same, so we can makes it a method like:
```
func (t *TAbleBuffer) appendRow(row map[string]column, encodeFn func(cols, item.RowID, t.colPerm) pair) error {
    cols, err := t.translateToDatum(row.Delete)
    if err != nil {
	return err
    }
    pair, err := encodeFn(cols, item.RowID, t.colPerm)
    if err != nil {
        return errors.Trace(err)
    }
    t.KvPairs = append(t.KvPairs, pair)
    t.Size++
}
```